File: ts-travel-plan-service/app/clients/travel_client.py
from __future__ import annotations

import httpx

from app.clients.base import Resolver
from app.domain.models import TripInfo, TripResponse

TRAVEL_SERVICE = "ts-travel-service"


class TravelClient:
    def __init__(self, resolver: Resolver, client: httpx.AsyncClient) -> None:
        self.resolver = resolver
        self.client = client

    async def left_trips(self, info: TripInfo) -> list[TripResponse]:
        travel_service_url = await self.resolver.resolve(TRAVEL_SERVICE)
        result = await self.client.post(
            travel_service_url + "/api/v1/travelservice/trips/left",
            json=info.model_dump(),
        )
        result.raise_for_status()
        return [TripResponse.model_validate(item) for item in result.json()["data"]]

File: ts-route-plan-service/internal/clients/travel_client.go
package clients

import (
	"context"
	"net/http"
	"net/url"

	"github.com/FudanSELab/train-ticket/ts-route-plan-service/internal/discovery"
	"github.com/FudanSELab/train-ticket/ts-route-plan-service/internal/domain"
)

const travelServiceName = "ts-travel-service"

type TravelClient struct{ transport transport }

func NewTravelClient(resolver discovery.Resolver, client *http.Client) *TravelClient {
	return &TravelClient{transport: newTransport(resolver, client)}
}

func (c *TravelClient) LeftTrips(ctx context.Context, info domain.TripInfo) ([]domain.TripResponse, error) {
	var response domain.Response[[]domain.TripResponse]
	err := c.transport.exchange(ctx, travelServiceName, http.MethodPost, "/api/v1/travelservice/trips/left", info, &response)
	return response.Data, err
}

func (c *TravelClient) RouteForTrip(ctx context.Context, tripID string) (*domain.Route, error) {
	var response domain.Response[*domain.Route]
	err := c.transport.exchange(ctx, travelServiceName, http.MethodGet, "/api/v1/travelservice/routes/"+url.PathEscape(tripID), nil, &response)
	return response.Data, err
}

func (c *TravelClient) TripsByRoutes(ctx context.Context, routeIDs []string) ([][]domain.Trip, error) {
	var response domain.Response[[][]domain.Trip]
	err := c.transport.exchange(ctx, travelServiceName, http.MethodPost, "/api/v1/travelservice/trips/routes", routeIDs, &response)
	return response.Data, err
}

func (c *TravelClient) TripDetail(ctx context.Context, info domain.TripAllDetailInfo) (*domain.TripAllDetail, error) {
	var response domain.Response[*domain.TripAllDetail]
	err := c.transport.exchange(ctx, travelServiceName, http.MethodPost, "/api/v1/travelservice/trip_detail", info, &response)
	return response.Data, err
}
